File: react_native_app/src/features/productos/domain/repositories/ProductoRepository.ts
import { Producto } from '../entities/Producto';

/**
 * Contrato del repositorio (capa `domain`). La capa `data` lo implementa
 * coordinando el datasource remoto (API) y el de memoria (RAM).
 */
export interface ProductoRepository {
  getProductos(forceRefresh?: boolean): Promise<Producto[]>;
  getProducto(id: number): Promise<Producto>;
  searchProductos(query: string): Promise<Producto[]>;
  createProducto(producto: Producto): Promise<Producto>;
  updateProducto(producto: Producto): Promise<Producto>;
  deleteProducto(id: number): Promise<void>;
}

File: flutter_app/lib/features/productos/domain/repositories/producto_repository.dart
import '../entities/producto.dart';

/// Contrato del repositorio (capa `domain`). La capa `data` lo implementa
/// coordinando el datasource remoto (API) y el de memoria (RAM).
abstract interface class ProductoRepository {
  Future<List<Producto>> getProductos({bool forceRefresh = false});
  Future<Producto> getProducto(int id);
  Future<List<Producto>> searchProductos(String query);
  Future<Producto> createProducto(Producto producto);
  Future<Producto> updateProducto(Producto producto);
  Future<void> deleteProducto(int id);
}
